File: lista/Python/questao15.py
"""
Crie uma classe de exceção personalizada SaldoInsuficienteException em Java e Python,
que seja lançada quando houver uma tentativa de saque superior ao saldo disponível na classe
ContaBancaria.
"""
class SaldoInsuficienteException(Exception):
    def __init__(self, mensagem):
        super().__init__(mensagem)

class ContaBancaria:
    def __init__(self, titular, saldo_inicial=0):
        self.titular = titular
        self._saldo = saldo_inicial

    def get_saldo(self):
        return self._saldo

    def depositar(self, valor):
        if valor > 0:
            self._saldo += valor
            print(f"Depositado: R${valor:.2f}. \nSaldo atual: R${self._saldo:.2f}.\n")
        else:
            print("Valor de depósito inválido.\n")

    def sacar(self, valor):
        if valor > self._saldo:
            raise SaldoInsuficienteException("Saldo insuficiente para realizar o saque.\n")
        elif valor <= 0:
            print("Valor de saque inválido.\n")
        else:
            self._saldo -= valor
            print(f"Sacado: R${valor:.2f}.\nSaldo atual: R${self._saldo:.2f}.\n")

def main():
    conta = ContaBancaria("João", 1000)

    print(f"Saldo inicial: R${conta.get_saldo():.2f}\n")

    try:
        conta.depositar(500)
        conta.sacar(300)
        conta.sacar(1500)
    except SaldoInsuficienteException as e:
        print(e)

if __name__ == "__main__":
    main()

File: lista/GO/questao15/questao15.go
package main

import (
	"fmt"
)

type SaldoInsuficienteError struct {
	Valor float64
}

func (e *SaldoInsuficienteError) Error() string {
	return fmt.Sprintf("Saldo insuficiente para saque de R$%.2f", e.Valor)
}

type ContaBancaria struct {
	Titular string
	Saldo   float64
}

func (c *ContaBancaria) Depositar(valor float64) {
	if valor > 0 {
		c.Saldo += valor
	} else {
		fmt.Println("Valor inválido.")
	}
}

func (c *ContaBancaria) Sacar(valor float64) error {
	if valor > c.Saldo {
		return &SaldoInsuficienteError{Valor: valor}
	}
	c.Saldo -= valor
	return nil
}

func (c *ContaBancaria) ExibirConta() {
	fmt.Printf("Titular: %s\n", c.Titular)
	fmt.Printf("Saldo: R$%.2f\n", c.Saldo)
}

func main() {
	conta := ContaBancaria{Titular: "João", Saldo: 1000}
	conta.ExibirConta()

	if err := conta.Sacar(300); err == nil {
		fmt.Println("Saque realizado com sucesso.")
	} else {
		fmt.Println("Erro:", err)
	}

	conta.ExibirConta()

	if err := conta.Sacar(1000); err != nil {
		fmt.Println("Erro:", err)
	}
}
